aesara/tensor/math_opt.py

```diff
@@ -3495,6 +3504,79 @@ def local_reciprocal_1_plus_exp(fgraph, node):
                         return out
 
 
+# x_lv, if any, is the logic variable version of some term x,
+# while x_at, if any, is the Aesara tensor version for the same.
+def distribute_over_add(in_lv, out_lv):
+    from cons import cons
+    from kanren import eq, fact, heado, tailo
+    from kanren.assoccomm import eq_assoccomm
+
+    # This does the optimization
+    # 1. (x + y + z) * A  = x * A + y * A + z * A
+    # 2. (x + y + z) / A  = x / A + y / A + z / A
+    A_lv, op_lv, add_term_lv, add_cdr_lv, mul_cdr_lv, add_flat_lv = lvars(6)
+    fact(commutative, at.mul)
+    fact(associative, at.mul)
+    fact(commutative, at.add)
+    fact(associative, at.add)
+    return lall(
+        # Make sure the input is a `at.mul` or `at.true_div`
+        lany(
+            eq(etuple(op_lv, A_lv, add_term_lv), in_lv),
+            eq(etuple(op_lv, add_term_lv, A_lv), in_lv),
+        ),
+        lany(eq(op_lv, at.mul), eq(op_lv, at.true_div)),
+        # Make sure the outer term being is an `add`
+        heado(at.add, add_term_lv),
+        # Flatten the associative pairings of `add` operations
+        assoc_flatten(add_term_lv, add_flat_lv),
+        # Get the flattened `add` arguments
+        tailo(add_cdr_lv, add_flat_lv),
+        # Add all the arguments and set the output
+        eq(cons(at.add, mul_cdr_lv), out_lv),
+        mapo(
+            (lambda x, y: eq_assoccomm(cons(op_lv, etuple(x, A_lv)), y)),
```

Comment:
In order to handle terms like `(+ A (* A D) ...)`, we could consider extending the `mapo`ed goal so that it handles two cases (e.g. with `conde`/`lany`): the current `(* A D)` product terms and  simply `A`.

For example, something like
```python
lambda x, y: conde([eq_assoccomm(cons(op_lv, etuple(x, A_lv)), y)], [eq(x, one_at), eq(y, A_lv)])
```
where `one_at` is something that unifies with any relevant form of `Constant(1)`.